internal/p2p/channel.go

```diff
@@ -259,17 +264,85 @@ func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
 }
 
 func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	// TODO: consider caching an iterator in the channel, or
+	// erroring if this gets called more than once.
+	//
+	// While it's safe to register a handler more than once, we
+	// could get into a dodgy situation where if you call receive
+	// more  than once, the subsequently messages won't be routed
+	// correctly.
 	iter := &ChannelIterator{
 		pipe: make(chan Envelope),
 	}
 
+	ch.host.SetStreamHandler(protocol.ID(ch.canonicalizedTopicName()), func(stream network.Stream) {
+		buf := bufio.NewReader(stream)
```

Comment:
`protoio.NewDelimitedReader` already handles this.

Reply:
it appears that we [create a bufio reader here](https://github.com/tendermint/tendermint/blob/master/internal/p2p/conn/connection.go#L169) and pass it to NewDelimtedReader [here](https://github.com/tendermint/tendermint/blob/master/internal/p2p/conn/connection.go#L471), and I was just following that lead.

Reply:
and the comment [here](https://github.com/tendermint/tendermint/blob/master/internal/libs/protoio/reader.go#L42-L46) suggests that we do need to buffer things.

Reply:
I looked at the implementation and it calls `bufio.NewReader` 🙂

Reply:
wait, really, I just looked at the code? Our `protoio` package appears to be foreked from gogoproto?

Reply:
oh god, I didn't realize that they'd *also* forked gogoproto?

Reply:
This is why we can't have nice things 😭